traitsui/tests/test_editor.py

```diff
@@ -0,0 +1,733 @@
+    auxiliary_event = Event
+
+    #: An auxiliary int we want to synchronize with a context value.
+    auxiliary_cv_int = Int(sync_value='from')
+
+    #: An auxiliary float we want to synchronize with a context value.
+    auxiliary_cv_float = Float
+
+    def init(self, parent):
+        self.control = FakeControl()
+        self.is_event = self.factory.is_event
+        if self.is_event:
+            self.control.on_trait_change(self.update_object, 'control_event')
+        else:
+            self.control.on_trait_change(self.update_object, 'control_value')
+
+    def dispose(self):
+        if self.is_event:
+            self.control.on_trait_change(self.update_object, 'control_event',
+                                         remove=True)
+        else:
+            self.control.on_trait_change(self.update_object, 'control_value',
+                                         remove=True)
+        super(StubEditor, self).dispose()
+
+    def update_editor(self):
+        if self.is_event:
+            self.control.control_event = True
+        else:
+            self.control.control_value = self.value
+
+    def update_object(self, new):
+        if self.control is not None:
+            if not self._no_update:
+                self._no_update = True
+                try:
+                    self.value = new
+                finally:
+                    self._no_update = False
+
+    def set_focus(self, parent):
+        pass
+
+
+class UserObject(HasTraits):
+
+    #: The value being edited.
+    user_value = Any('test')
+
+    #: An auxiliary user value
+    user_auxiliary = Any(10)
+
+    #: An list user value
+    user_list = List(['one', 'two', 'three'])
+
+    #: An event user value
+    user_event = Event
+
+
+@unittest.skipIf(is_current_backend_null(),
+                 'Editor tests require non-null backend')
+class TestEditor(UnittestTools, unittest.TestCase):
+
+    def create_editor(self, context=None, object_name='object',
+                      name='user_value', factory=None, is_event=False):
+        if context is None:
+            user_object = UserObject()
+            context = {'object': user_object}
+        elif '.' in object_name:
+            context_name, xname = object_name.split('.', 1)
+            context_object = context[context_name]
+            user_object = xgetattr(context_object, xname)
+        else:
+            user_object = context[object_name]
+        ui = UI(
+            context=context,
+            handler=default_handler(),
+        )
+
+        if factory is None:
+            factory = StubEditorFactory()
+        factory.is_event = is_event
+
+        editor = StubEditor(
+            parent=None,
+            ui=ui,
+            object_name=object_name,
+            name=name,
+            factory=factory,
+            object=user_object,
+        )
+        return editor
+
+    def test_lifecycle(self):
+        editor = self.create_editor()
```

Comment:
Should we have some kind of try/finally or context-manager setup to make sure that editors always get disposed in these tests, even if one of the test assertions has failed?

Reply:
In this particular test suite it doesn't matter because there is no actual toolkit control being created - everything (including the `FakeControl`) is pure Python.